src/bson_array.c

```diff
@@ -207,6 +208,144 @@ BsonArray bson_array_from_bytes(uint8_t *data) {
   return array;
 }
 
+size_t bson_array_from_bytes_len(BsonArray *output, const uint8_t *data, size_t data_size) {
+  const uint8_t *current = data;
+  size_t remain_bytes = data_size;
+  int32_t size = 0;
+  bool parse_error = false;
+  size_t ret;
+
+  if (output == NULL || data == NULL) {
+    return 0;
+  }
+  ret = read_int32_le_len(&size, &current, &remain_bytes);
+  if (ret == 0) {
+    return 0;
+  }
+  if (size > data_size) {
+    printf("Unexpected array length %i, data is only %i bytes\n", (int)size, (int)data_size);
+    return 0;
+  }
+
+  uint8_t type;
+  ret = read_byte_len(&type, &current, &remain_bytes);
+  if (ret == 0) {
+    return 0;
+  }
+
+  BsonArray array;
+  bson_array_initialize(&array, 10);
+
+  while (type != DOCUMENT_END) {
+    char *key = NULL;
+    ret = read_string_len(&key, &current, &remain_bytes);
+    if (ret == 0) {
+      parse_error = true;
+      break;
+    }
+    ret = 0;
+
+    switch ((element_type)type) {
+      case TYPE_DOCUMENT: {
+        BsonObject obj;
+        ret = bson_object_from_bytes_len(&obj, current, remain_bytes);
+        if (ret > 0) {
+          bson_array_add_object(&array, &obj);
+          current += ret;
+          remain_bytes -= ret;
+        }
+        break;
+      }
+      case TYPE_ARRAY: {
+        BsonArray subArray;
+        ret = bson_array_from_bytes_len(&subArray, current, remain_bytes);
+        if (ret > 0) {
+          bson_array_add_array(&array, &subArray);
+          current += ret;
+          remain_bytes -= ret;
+        }
+        break;
+      }
+      case TYPE_INT32: {
+        int32_t value = 0;
+        ret = read_int32_le_len(&value, &current, &remain_bytes);
```

Comment:
I don't believe a `len` variant is necessary for the primitive types, as you can just do a length check before reading since they have a fixed length.